subsys/net/ip/tcp.c

```diff
@@ -2379,25 +2379,32 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
 				th_seq(th) == conn->ack)) {
 			k_work_cancel_delayable(&conn->establish_timer);
 			tcp_send_timer_cancel(conn);
-			next = TCP_ESTABLISHED;
 			tcp_conn_ref(conn);
 			net_context_set_state(conn->context,
 					      NET_CONTEXT_CONNECTED);
 
-			if (conn->accepted_conn) {
-				if (conn->accepted_conn->accept_cb) {
-					conn->accepted_conn->accept_cb(
-						conn->context,
-						&conn->accepted_conn->context->remote,
-						sizeof(struct sockaddr), 0,
-						conn->accepted_conn->context);
-				}
-
-				/* Make sure the accept_cb is only called once.
+			if (conn->accepted_conn == NULL ||
+			    conn->accepted_conn->accept_cb == NULL) {
+				/* In case of no accept_cb registered,
+				 * application will not take ownership of the
+				 * connection. To prevent connection leak, unref
+				 * the TCP context and put the connection into
+				 * active close (TCP_FIN_WAIT_1).
 				 */
-				conn->accepted_conn = NULL;
+				net_tcp_put(conn->context);
+				break;
 			}
 
+			conn->accepted_conn->accept_cb(
+				conn->context,
+				&conn->accepted_conn->context->remote,
+				sizeof(struct sockaddr), 0,
+				conn->accepted_conn->context);
+
+			/* Make sure the accept_cb is only called once. */
+			conn->accepted_conn = NULL;
```

Comment:
Just wondering if there could be race here when calling accept_cb and setting it to null. One safer way would be to set the accepted_conn to null before calling it.

Reply:
Hmm possibly. I've added a local copy of `accept_cb` pointer and call it now after the `accepted_conn` is cleared. Please check if it matches what you thought of.